### DISCUSSION TITLE
DataHub get all documents

### DISCUSSION THREAD

**Discussion by snmabaur** [pimcore/pimcore#13391] [Q&A / Support]
Hi

is there a way to get all documents through graphql? In the documentation I just see the `getDocument(id:int)` query but nothing to get all pages or even all the id's of pages there are... 

For objects there is the `getObjectListing` query but it's missing for the documents... 

Basically I'd like to get the whole page structure for a navigation 

Thanks for any help
Cheers

**Comment by mattamon** [comment]
Hey @snmabaur,

the only way I can think of, would be like union operations.

You could do something like this, e.g. getting the home document and then specify every single document type.
``` graphql
query {
  getDocument(id: 1) {
    ... on document_page {
      children {
        ... on document_page {
          id
          title
          fullpath
          type
          children {
            ... on document_snippet {
              id
              type
            }
            ... on document_email {
              id
              type
            }
          }
        }
      }
    }
  }
}
```
Unfortunately I see no easier way of doing this atm. 
I agree that a document listing would be helpful if you want e.g. the navigation.
For further information I would refer you to:
https://pimcore.com/docs/data-hub/current/GraphQL/Query/Document_Queries.html
https://pimcore.com/docs/data-hub/current/GraphQL/Mutation/Document_Mutations.html

**Reply by snmabaur** [reply]
Thanks, but with this solution I can't do it full dynamic since I don't know how many child pages of child pages will be created, it would really be nice if we could get the whole navigation throuh GraphQL

For the momentn I'm building my own API to endpoint

So far I wrote a recursive function to get all pages with the child pages and it works quite well

```
private function buildNav(int $parentId): array {
    $listing = Document::getList([
        'unpublished' => true,
        'condition' => "`parentId` = ${parentId}",
        'orderKey' => ['key', 'published'],
        'order' => 'desc',
    ]);
    $documents = $listing->getDocuments();

    $children = array();
    foreach ($documents as $document) {
        $children[] = [
            'id' => $document->getId(),
            'path' => $document->getPath() . $document->getKey(),
            'seo_title' => $document->getTitle(),
            'nav_title' => $document->getProperty('navigation_title'),
            'children' => $this->buildNav($document->getId(), $lang),
        ];
    }
    return $children;
}
```


Is there a better way to get the whole nav tree in pimcore? I know that in the templates a can use the build_nav twig helper function but somehow I can't find anything to get the whole structure in my controller

Thanks a lot

**Reply by mattamon** [reply]
Well I agree that this would be a good functionality for GraphQL but at the moment this seems not to be possible.

For the controller, you can always inject the navigation builder into your controller to get the whole navigation tree.

```
use Pimcore\Navigation\Builder;

public function defaultAction(Request $request, Builder $builder): Response
{
     $navigation = $builder->getNavigation([]);
     return $this->render('default/default.html.twig');
}
```

The same function is used in the twig extension.
https://github.com/pimcore/pimcore/blob/10.5/lib/Twig/Extension/NavigationExtension.php#L45
https://github.com/pimcore/pimcore/blob/10.5/lib/Twig/Extension/Templating/Navigation.php#L73
https://github.com/pimcore/pimcore/blob/10.5/lib/Navigation/Builder.php#L129

Please be aware that there will be breaking changes in Pimcore 11 for this function
`'Calling Pimcore\Navigation\Builder::getNavigation() using extra arguments is deprecated and will be removed in Pimcore 11.         Instead, specify the arguments as an array'`
